src/python/pants/engine/fs.py

```diff
@@ -63,18 +63,39 @@ class PathGlobs:
   exclude: Tuple[str, ...]
   glob_match_error_behavior: GlobMatchErrorBehavior
   conjunction: GlobExpansionConjunction
+  description_of_origin: str
 
   def __init__(
     self,
     include: Iterable[str],
     exclude: Iterable[str] = (),
     glob_match_error_behavior: GlobMatchErrorBehavior = GlobMatchErrorBehavior.ignore,
-    conjunction: GlobExpansionConjunction = GlobExpansionConjunction.any_match
+    conjunction: GlobExpansionConjunction = GlobExpansionConjunction.any_match,
+    description_of_origin: Optional[str] = None,
   ) -> None:
+    """
+    :param include: globs to match, e.g. "foo.txt" or "**/*.txt"
+    :param exclude: globs to exclude, in the same style as the args to `include`
+    :param glob_match_error_behavior: whether to warn or error upon match failures
+    :param conjunction: whether all `include`s must match or only at least one must match
+    :param description_of_origin: a human-friendly description of where this PathGlobs request is
+                                  coming from, used to improve the error message for unmatched
+                                  globs. For example, this might be
+                                  "src/python/pants/util:strutil (line 84)".
+    """
     self.include = tuple(include)
     self.exclude = tuple(exclude)
     self.glob_match_error_behavior = glob_match_error_behavior
     self.conjunction = conjunction
+    self.description_of_origin = description_of_origin or ""
+    self.__post_init__()
+
+  def __post_init__(self) -> None:
+    if not self.description_of_origin and self.glob_match_error_behavior != GlobMatchErrorBehavior.ignore:
```

Comment:
We have a lot of `PathGlobs` uses that use the default `ignore` and it seems for good reason, that it's safe to not find the globs in those use cases. 

When set to `ignore`, it's impossible for the engine to display an error about unmatched globs. So, we only need to ensure that `description_of_origin` is populated when the user can encounter an error.